mautic/mautic
# Pipedrive full sync

[//]: # ( Please answer the following questions: )

| Q  | A
| --- | ---
| Bug fix? | 
| New feature? | y
| Automated tests included? |
| Related user documentation PR URL | https://github.com/mautic/documentation/pull/241
| Related developer documentation PR URL | 
| Issues addressed (#s or URLs) | https://github.com/mautic/mautic/issues/5164
| BC breaks? | 
| Deprecations? | 

[//]: # ( Note that all new features should have a related user and/or developer documentation PR in their respective repositories. )

This bugfix/feature has been sponsored by @Webmecanik

[//]: # ( Required: )
#### Description:
PR added missed sync between Mautic and Pipedrive
From this request: https://github.com/mautic/mautic/issues/5164

- ignore anonymouse contact if you call  `mautic:integration:pipedrive:push`
- pretend duplicate contact create if your contacts already exist but wasn't matched before  (if you call mautic:integration:pipedrive:fetch and mautic:integration:pipedrive:push)
- added listener to update lead in Mautic and update contact on Pipdrive realtime

Now plugin should cover every scenario. I hope :)

Please test also  commands:

mautic:integration:pipedrive:fetch
mautic:integration:pipedrive:push

#### Steps to test this PR:
1.  Setup https://mautic.org/docs/en/plugins/pipedrive.html (for test trial is useful)
2.  Create contact in Pipedrive and see if contact was created in Mauitc
3.  Edit contact in Pipedrive and see if contact was modified in Mautic
4. Create contact in Mautic and see if contact was created  in Pipedrive
5. Modify contact in Mautic and see if contact was modifeid in Pipedrive
6.  Disable Pipedrive plugin
7. Create new contact in Mautic and another contact in Pipdedrive. 
8. Enable Pipedrive plugin
9. Go to Mautic and edit contact created while plugin was disabled (step 6 ). Contact should  be created in Pipedrive
10. Go to Pipedrive and edit contact create while plugin was disabled (step 6).  Contact should be created in Mautic
11.  Disable Pipedrive plugin
12. Go to Mautic and Pipedrive and create new contact (same for both systems)
13. Enable Pipedrive plugin
14. Try edit contact in Mautic, should by updated also on Pipedrive (not created like before)
15. Then go to database and remove some lines from integration_entity
16: Try call  `php app/console mautic:integration:pipedrive:fetch` and  `php app/console mautic:integration:pipedrive:push` 
17. See if contacts was just merged, not created  like before

This PR also fixes **Pipedrive cleanUpFields issue**  https://github.com/mautic/mautic/pull/5746 

[//]: # ( As applicable: )
#### Steps to reproduce the bug:

Require setup Pipedrive with trial account https://mautic.org/docs/en/plugins/pipedrive.html 

1. Create a custom field in Pipedrive
2. Map it in integration mapping tab
3. Delete the custom field in Pipedrive
4. Notice that in the integration_settings table the mapping with the deleted field is still existing. You can try to open the mapping and save & close but in DB the value stays
5. Push a contact and notice error "the field XXX is not existing"

#### Steps to test this PR:
1.  Setup Pipedrive first
2.  Then go to plugin and see if everything work like before
3. Then repeat steps to reproduce
4: Works If fields are clean up after save settings

This PR also add new **--restart** param to commands  from this closed PR https://github.com/mautic/mautic/pull/6457
We notice some clients are crazy and disable plugin and delete data, and then can sync all data between Mautic and Pipedrive. We add new param --restart to both commands. This parameter before sync data remove all data from integration_entity. Then plugin try find all contacts and companies and create integration_entity again. 

[//]: # ( As applicable: )
#### Steps to test:

1. After all test and sync data. Disable plugin
2. Then go to Pipedrive and remove all companies
3. Then enable plugin 
4. Run push from Mautic with restart param `php app/console mautic:integration:pipedrive:push – restart`
5. All companies should created and new integration_entity items should created without duplicates

Usage:

```
php app/console mautic:integration:pipedrive:fetch --restart
php app/console mautic:integration:pipedrive:push – restart
```
